doc/developers/contributing.rst

```diff
@@ -734,7 +736,8 @@ In general have the following in mind:
     5. Specify ``dataframe`` when "frame-like" features are being used, such
        as the column names.
     6. When specifying the data type of a list, use ``of`` as a delimiter:
-       ``list of int``.
+       ``list of int``. If the data type was also mentioned as possible output,
+       you can use ``list of thereof`` to be more concise.
```

Comment:
Cross reference: https://github.com/scikit-learn/scikit-learn/pull/17306#discussion_r432836089

Reply:
It should be "~of~ thereof" I believe.

And I think we should give a more concrete example:

> For example if a parameter supports ints, floats, lists of ints, and list of float, you can use "int, floats, or list thereof".

Also I don't think using "thereof" is warranted when it refers to only one thing: "int or list thereof"  should probably just be "list or list of int"


Then again, not a native speaker...

Reply:
I see what you mean now. As a native speaker, I feel that this phrasing will be confusing.

Reply:
isn't it correct English?
[Jax](https://jax.readthedocs.io/en/latest/jax.html#jax.vmap) uses it

Reply:
In any case we can also suggest "... or list of these"

Since "of these" / "of that" is precisely what "thereof" means

Reply:
It is correct english, but I have not seen it commonly used. For example:

> ndarray of shape (n_samples, n_classes) or list of thereof

The `thereof` is referring to `ndarray`. I think this type of "going back through the sentence" and having to think about what `thereof` is referring will be difficult for some.

We can try it out and see if it leads to any confusion.

XREF: https://english.stackexchange.com/questions/27654/why-do-we-use-the-word-thereof

Reply:
So let's try out. It is probably as confusing as the difference between an ndarray and an array-like at the difference that is documented in an English dictionary instead of our Glossary :)

Reply:
@glemaitre could you apply the changes proposed in https://github.com/scikit-learn/scikit-learn/pull/18243/files#r475250655 ;)

Reply:
Good point. It makes more sense.